vsintegration/src/FSharp.Editor/QuickInfo/QuickInfoProvider.fs

```diff
@@ -53,45 +53,70 @@ type internal FSharpQuickInfoProvider
         checkerProvider: FSharpCheckerProvider,
         projectInfoManager: ProjectInfoManager,
         typeMap: Shared.Utilities.ClassificationTypeMap,
+        gotoDefinitionService:FSharpGoToDefinitionService,
         glyphService: IGlyphService
     ) =
 
-    let fragment(content: Layout.TaggedText seq, typemap: ClassificationTypeMap, thisDoc: Document) =
+    let fragment (content: Layout.TaggedText seq, typemap: ClassificationTypeMap, initialDoc: Document) : IDeferredQuickInfoContent =
 
-        let workspace = thisDoc.Project.Solution.Workspace
+        let workspace = initialDoc.Project.Solution.Workspace
         let solution = workspace.CurrentSolution
 
-        let documentId (range: range) =
+        /// Retrieve the DocumentId from the workspace for the range's file
+        let docIdOfRange (range: range) =
             let filePath = System.IO.Path.GetFullPathSafe range.FileName
-            let projectOf (id : DocumentId) = solution.GetDocument(id).Project
 
-            //The same file may be present in many projects. We choose one from current or referenced project.
-            let rec matchingDoc = function
-            | [] -> None
-            | id::_ when projectOf id = thisDoc.Project || IsScript thisDoc.FilePath -> Some id
-            | id::tail -> 
-                if (projectOf id).GetDependentProjects() |> Seq.contains thisDoc.Project then Some id
-                else matchingDoc tail
-            solution.GetDocumentIdsWithFilePath(filePath) |> List.ofSeq |> matchingDoc
+            // The same file may be present in many projects. We choose one from current or referenced project.
+            let candidates = solution.GetDocumentIdsWithFilePath filePath
+            match candidates.Length with 
+            | 0 -> None 
+            | 1 -> 
```

Comment:
The change I noticed is that with my version I can navigate from, say FSharp.Editor to FSharp.Core.
The important part was the recursion, I think.